be/src/exprs/aggregate_functions.cpp

```diff
@@ -2534,6 +2535,97 @@ IntVal AggregateFunctions::window_funnel_finalize(FunctionContext* ctx, const St
     return doris_udf::IntVal(val);
 }
 
+// Refer to AggregateFunctionRetention.h in https://github.com/ClickHouse/ClickHouse.git
+struct RetentionState {
+    static constexpr size_t max_events = 32;
+    std::bitset<max_events> events;
+
+    RetentionState() {}
+
+    void add(int event) { events.set(event); }
+
+    void merge(RetentionState* other) { events |= other->events; }
+
+    BooleanVal* getRetentionData() {
+        static BooleanVal data_to[max_events];
+        int current_offset = 0;
+
+        const bool first_flag = events.test(0);
+        data_to[current_offset] = first_flag;
+        ++current_offset;
+
+        for (size_t i = 1; i < events.size(); ++i) {
+            data_to[current_offset] = (first_flag && events.test(i));
+            ++current_offset;
+        }
+        return data_to;
+    }
+
+    int64_t serizlized_size() { return sizeof(events); }
+
+    void serialize(uint8_t* buf) {
+        memcpy(buf, &events, sizeof(events));
```

Comment:
bitset is a container, memcpy should not be used for serialized.

Reply:
BTW, you can just remove non-vectorized implementation.  Because we will remove non-vectorized code in the future, maybe in version 1.3 or 1.4.

Reply:
will do.